### DISCUSSION TITLE
Alternative to $this->render(); in v2

### DISCUSSION THREAD

**Discussion by qwasyx0** [rappasoft/laravel-livewire-tables#766] [Q&A] [answered]
Hello, 

I'm strugglign with refreshing the table after external event update. I'm always one step behind (after next external event update the previous $this->externalId is executed in builder()). Is there a way to run  the basic query after the $this->externalId is updated?

Here is the code:
```
 public $externalId;

    protected $listeners = [
        'updateExternalId'
    ];

    public function updateExternalId($value)
    {
        $this->externalId = $value;

        // trying to refresh does nothing, according to documentation this should do the job
        $this->emit('refreshDatatable');

        // this chains the query so there are two wheres
        $query = $this->baseQuery()->getQuery();
        $this->executeQuery($query->where('external_id', $this->externalId));
    }

    public function builder(): Builder
    {
        $query = Meridlo::query();

        if ($this->externalId != null) {
            $query = $query->where('external_id', $this->externalId);
        }

        return $query;
    }
```

**Comment by AndMe2** [comment]
Hi! Any news?
We got the same problem but with a little different approach, we send the whole query via an $emit and do an eval() on it in Builder. Works like a charm, but as you point out the variable is updated after the Builder() is run. So the second time I click the link the query is correct because it was updated the first time, but after builder. So it is one step behind in the flow as qwasyx0  point out and every link has to be clicked twice to get the correct query result right now.
Maby some config option for builder()? or able to send value to builder()

**Comment by eumakue92** [comment, marked as answer]
I think I just found a way. I had a problem that is related to this.

If you want to execute the builder just after something, you can use the following:

`$this->setBuilder($this->builder());`

I think this shouldn't be the way, but it works to me better than _refreshDatatable_, which just loaded an empty table.

**Comment by AndMe2** [comment]
On the same path, we moved $this->setBuilder($this->builder()); and $this->setColumns(); to render() from booted() for now.

**Comment by jacksontong** [comment]
I think this is a very serious problem, I was struggling with this issue until I found this post, I expect the data would refresh when I call `$this->emitSelf('refreshDatatable');` but it's always one step behind.

**Comment by michel2103** [comment]
I'm facing the same problem. I'm filtering between two dates and do some calcs, but I'm one step behind in updates. 
When I emit refreshDatatable the component is refreshing every time and I don't know how to stop it. The refresh doesn't allow me to select a new date filter, because the component refreshes and the date picker disappears.
